### PR TITLE
[WIP] Draw line on x

### PR DESCRIPTION
**Pull request by Cogitri**
This still needs scrolling to be fixed (right now the line just appears after 80chars and if you scroll it stays in the same position).

### CODE REVIEW

**Comment by Cogitri** [comment]
Alright, this works now, but I'm not quite sure on the implementation. We could either do the following:

screenshot from 2019-03-08 16-14-19

So just one line which acts as indicator for the (soft) char limit

Other editors do it like this:

screenshot from 2019-03-08 16-12-26

So they draw one rectangle for the OK-area and one rectangle for the This-is-too-wide-area

Opinion, @LiHRaM?

**Comment by LiHRaM** [comment]
The second one seems better to me. It seems like a more natural transition.

**Comment by Cogitri** [comment]
Alright, thanks :)

**Comment by Cogitri** [comment]
Looks like this now:

screenshot from 2019-03-08 18-37-19

**Comment by LiHRaM** [comment]
Looks good - do you think it would make sense to have the middle portion lighter and the two sides darker, though? I guess that's just what I expected based on the second example.

**Comment by Cogitri** [comment]
> Looks good - do you think it would make sense to have the middle portion lighter and the two sides darker, though?

Ah, that highly depends on what theme is chosen.

File: src/edit_view.rs [hunk 1]
```diff
@@ -685,10 +683,28 @@ impl EditView {
...
-        set_source_color(cr, theme.background);
-        cr.rectangle(0.0, 0.0, f64::from(da_width), f64::from(da_height));
-        cr.fill();
+        // Draw a line at x chars
+        if get_draw_right_margin() {
+            let until_margin_width = self.edit_font.font_width * get_column_right_margin() as f64;
+            // Draw editing background
+            set_source_color(cr, theme.background);
+            cr.rectangle(0.0, 0.0, until_margin_width, f64::from(da_height));
+            cr.fill();
```
Comment (Cogitri) [inline src/edit_view.rs]: We use whatever Xi tells us the background is here (this is the darker part in the screenshot)

File: src/edit_view.rs [hunk 2]
```diff
@@ -685,10 +683,28 @@ impl EditView {
...
+
+            set_source_color(cr, theme.gutter);
+            cr.rectangle(
```
Comment (Cogitri) [inline src/edit_view.rs]: We draw with the gutter color for the linecount and for the right hand margin. This has the not so nice side effect that upon scrolling to the right the right hand margin and the linecount are next to each other and both have the same color.

**Comment by LiHRaM** [comment]
This makes me feel like maybe xi theming options don't account for this sort of design, maybe going back to option 1 makes the most sense, then?

**Comment by Cogitri** [comment]
I've asked in Xi's zulipchat, let's see. FWIW, do you think adding a black line to the rightmost of the linecount to limit it would be a good idea?

**Comment by LiHRaM** [comment]
> I've asked in Xi's zulipchat, let's see. FWIW, do you think adding a black line to the rightmost of the linecount to limit it would be a good idea?

Honestly, I think maybe we should try playing around with spacing a little more. A lot can be achieved with the appropriate padding. (I take inspiration from VSCode mostly, which I think is very well designed.)

**Comment by Cogitri** [comment]
Hm, so instead of having a different colour for them you'd suggest to just leave more space between the linecount and the editing area?

**Comment by LiHRaM** [comment]
Exactly. It works well for other editors at least.

**Comment by Cogitri** [comment]
Hmm..will look into that

**Comment by Cogitri** [comment]
It looks like this now:

Screenshot from 2019-03-10 12-48-47

The whole theming situation with Xi-Editor is little mediocre, how about we provide an option to just use whatever the GTK theme dictates (I think some other GTK editors have similar options - not quite sure how to implement it though, will look into it if you think it makes sense). We could also add our own themes, that'd most likely be easier for now.

**Comment by LiHRaM** [comment]
I think supporting both Xi and GTK themes makes sense, but it might 
also be too much work. Your second option of having our own themes also 
seems like a big job - maybe we could support custom theming options 
within the framework of xi theming, and then PR to the main repo for 
the ones we really like?

**Comment by Cogitri** [comment]
> maybe we could support custom theming options within the framework of xi theming, and then PR to the main repo for the ones we really like?

Yes, Xi does plan to rework the theming, but there's no schedule for that yet. We could just hardcode some `Theme`s into gxi for the time being, right now only Solarized* and Inspired GitHub are usable IMHO

**Comment by Cogitri** [comment]
I really do prefer the current look to just drawing a line (thanks with helping me with that decision), but it's not visible in InspiredGithub (the standard theme) :c

I'm kind of inclined to just work around this for InspiredGithub while we wait for the theming situation to get better at Xi's side.

File: meson_post_install.sh [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    schemadir="$MESON_INSTALL_PREFIX/share/glib-2.0/schemas"
+    icondir="$MESON_INSTALL_PREFIX/share/icons/hicolor"
+
+    printf "%s\\n" "Compiling GSettings schemas in dir $schemadir"
+    glib-compile-schemas $schemadir
+
+    printf "%s\\n" "Updating GTK icon cache in dir $icondir"
+    gtk-update-icon-cache -f -t $icondir
+
+    printf "%s\\n" "Updating desktop MIME database"
```
Comment (maxice8) [inline meson_post_install.sh]: (MIME)

File: meson_post_install.sh [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    update-desktop-database
+fi
```
Comment (maxice8) [inline meson_post_install.sh]: newline at the end ?

File: data/com.github.Cogitri.gxi.gschema.xml [hunk 5]
```diff
@@ -20,6 +20,21 @@
...
+            <description>
+                Draw a right hand margin to limit the text width
+            </description>
+        </key>
+
+        <key name="column-right-margin" type="u">
+            <default>80</default>
+            <summary>At what column to display the margin</summary>
+            <description>
+                Sets the column at what the margin should be drawn
```
Comment (Cogitri) [inline data/com.github.Cogitri.gxi.gschema.xml]: At which

File: data/com.github.Cogitri.gxi.gschema.xml [hunk 6]
```diff
@@ -20,6 +20,21 @@
...
+            </description>
+        </key>
     </schema>
 
 </schemalist>
```
Comment (Cogitri) [inline data/com.github.Cogitri.gxi.gschema.xml]: Add a newline

File: src/edit_view.rs [hunk 7]
```diff
@@ -762,34 +779,35 @@ impl EditView {
...
         // Draw linecount background
-        set_source_color(cr, theme.gutter);
+        set_source_color(cr, theme.background);
         cr.rectangle(0.0, 0.0, linecount_width, f64::from(linecount_height));
         cr.fill();
 
-        set_source_color(cr, theme.foreground);
         //FIXME: Xi sends us the 'ln' (logical linenumber) param for this, but that isn't updated on every draw!
         let mut current_line = first_line;
+        let center_diff = (self.edit_font.font_height - self.interface_font.font_height) / 2.0;
```
Comment (Cogitri) [inline src/edit_view.rs]: Make sure this is positive

**Comment by Cogitri** [comment]
This looms very similar to what gedit does now and simply uses a darker version of the background's color

**Pull request merged**